### PR TITLE
add max number of notices show

### PR DESCRIPTION
**Pull request by jzhangs**
To help fix https://github.com/ant-design/ant-design/issues/9198 , add a maxItems prop to set max number of notices show in current page. Update last item using same key if limit exceeded to avoid occasional flicker.

### CODE REVIEW

**Comment by jzhangs** [comment]
@nikogu please help review this

File: README.md [hunk 1]
```diff
@@ -145,6 +145,12 @@ props details:
           <td> { right: '50%' } </td>
           <td>additional style for single notice node.</td>
         </tr>
+        <tr>
+          <td>maxItems</td>
```
[the remark this answers is not in the document]
Comment (jzhangs) [reply README.md]: prefer `maxCount`

File: src/Notification.jsx [hunk 2]
```diff
@@ -28,6 +29,7 @@ class Notification extends Component {
       top: 65,
       left: '50%',
     },
+    maxItems: 99,
```
[the remark this answers is not in the document]
Comment (jzhangs) [reply src/Notification.jsx]: ok, will update this

File: src/Notification.jsx [hunk 3]
```diff
@@ -45,15 +47,21 @@ class Notification extends Component {
...
+      const updatedNotices = notices.concat();
+      if (noticeIndex !== -1) {
         updatedNotices.splice(noticeIndex, 1, notice);
+      } else {
+        if (notices.length < maxItems) {
+          updatedNotices.push(notice);
+        } else {
+          notice.key = updatedNotices[maxItems - 1].key;
+          notice.update = true;
+          updatedNotices[maxItems - 1] = notice;
```
[the remark this answers is not in the document]
Comment (jzhangs) [reply src/Notification.jsx]: Yes, and tried. Here's the problem: If use `remove top add bottom` mechnism as above, and say when `maxCount === 1`,  two notices will appear in a short time when unmounting previous one and mounting lastest. Personally that's not acceptable and I can't find a better solution without touch current anim design. 

Anyway, if `maxCount` is only used as 1 or unlimited, it looks fine, though make it FIFO is more straightforward.

[the remark this answers is not in the document]
Comment (nikogu) [reply src/Notification.jsx]: if maxCount === 1 then updatedNotices.update(notice)

it's better ? merge 2 solution done different scene.

[the remark this answers is not in the document]
Comment (jzhangs) [reply src/Notification.jsx]: Not that easy, if exceed `maxCount`, there's always chance that `maxCount + 1` notices would show.

[the remark this answers is not in the document]
Comment (nikogu) [reply src/Notification.jsx]: @jzhangs i think your are right, but in your way, every replace next one can solve this problem.

```js
  add = notice => {
    const key = (notice.key = notice.key || getUuid());
    const { maxItems } = this.props;
    this.setState(previousState => {
      const notices = previousState.notices;
      const noticeIndex = notices.map(v => v.key).indexOf(key);
      let updatedNotices = notices.concat();
      if (noticeIndex !== -1) {
        updatedNotices.splice(noticeIndex, 1, notice);
      } else {
        if (notices.length < maxItems) {
          updatedNotices.push(notice);
        } else {
          // save key for new refs
          const keys = updatedNotices.map(notice => notice.key);
          updatedNotices.push(notice); 
          updatedNotices.shift();
          keys.forEach((key, i) => {
            updatedNotices[i].update = true;
            updatedNotices[i].key = key;
          })
        }
      }
      return {
        notices: updatedNotices
      };
    });
  };

```

**Comment by jzhangs** [comment]
@nikogu Updated .. It's sad that there's some` 'controlled close'` mode in rc-notification that use notice.key outside to close notice, which means we can't simply update its value. And still, reuse 'key' as this is some hack that not good ..

Current patch is fine except if you set `maxCount: 3` , open localhost:8000/examples/simple.html and click `'simple show'` and `'updadable'` several times. Hardcoded key and setInterval make it a mess ..

File: src/Notification.jsx [hunk 4]
```diff
@@ -45,15 +46,22 @@ class Notification extends Component {
...
-        updatedNotices = notices.concat();
+      const updatedNotices = notices.concat();
+      if (noticeIndex !== -1) {
         updatedNotices.splice(noticeIndex, 1, notice);
+      } else {
+        if (maxCount && notices.length >= maxCount) {
+          // XXX, use key of first item to update new added (let React to move exsiting
+          // instead of remove and mount). Same key was used before for both a) external
+          // manual control and b) internal react 'key' prop , which is not that good.
+          notice.updateKey = updatedNotices[0].updateKey || updatedNotices[0].key;
```
[the remark this answers is not in the document]
Comment (jzhangs) [reply src/Notification.jsx]: It's an ugly hack.. as in [1], `notice.key` is used outside to manual close notice with 'duration:null', which means we can't modify it directly.

```
function close(key) {
  notification.removeNotice(key);
}

function manualClose() {
  const key = Date.now();
  notification.notice({
    content: <div>
      <p>click below button to close</p>
      <button onClick={close.bind(null, key)}>close</button>
    </div>,
    key,
    duration: null,
  });
}
```

[1] https://github.com/react-component/notification/blob/master/examples/simple.js#L37

[the remark this answers is not in the document]
Comment (jzhangs) [reply src/Notification.jsx]: Maybe a better name, `reusedKey` ?

[the remark this answers is not in the document]
Comment (nikogu) [reply src/Notification.jsx]: i think it's all right~

**Pull request merged**